An RFC for SpanContext, References, as well as changes to carriers

bensigelman opened the pull request:
I would love feedback about the API here... once we converge, we can decide what to do about the rest of this (github) repo (the impl helpers, etc).

Please look at https://github.com/opentracing/opentracing-java/issues/31 and https://github.com/opentracing/opentracing.github.io/issues/99 for background and motivation.

I'll add a few PR comments inline as well.

bensigelman commented on opentracing-api/src/main/java/io/opentracing/BuiltinFormats.java:
```diff
@@ -0,0 +1,44 @@
+ * @see io.opentracing.Tracer#inject(SpanContext, Format, Object)
+ * @see io.opentracing.Tracer#extract(Format, Object)
+ */
+public class BuiltinFormats<R, W> implements Format<R, W> {
```
@yurishkuro I went with `BuiltinFormats` instead of `Formats` -- seemed to better emphasize that non-builtin extensions were feasible.

bensigelman commented on opentracing-api/src/main/java/io/opentracing/BuiltinFormats.java:
```diff
@@ -0,0 +1,44 @@
+    /**
+     * TEXT_MAP carriers support arbitrary string:string mapping with no special escaping needed on either keys or values.
+     *
+     * The backing store may (or may not) be shared with other non-OpenTracing writers and readers, so it's important for Tracer implementations to use a prefix to identify their own entries in the text map.
```
what do we think about this for non-HTTP TextMap?

bensigelman commented on opentracing-api/src/main/java/io/opentracing/Format.java:
```diff
@@ -11,9 +11,16 @@
+ * @see io.opentracing.Tracer#inject(SpanContext, Format, Object)
+ * @see io.opentracing.Tracer#extract(Format, Object)
+ */
+public interface Format<R, W> {
```
I guess the builtin formats could also just be public static members inside here...

```
public interface Format<R, W> {
    public final static Format<String, String> TextMap = new Builtins<>();

    class Builtins<R, W> implements Format<R, W> {
    }
}
```

Thoughts?

adriancole replied:
the Builtins thing seems awkward to me. why not just inline the three constants here?

bensigelman commented on opentracing-api/src/main/java/io/opentracing/Reference.java:
```diff
@@ -0,0 +1,65 @@
+    private SpanContext referee;
+
+    public enum Type {
+        CHILD_OF,
```
I guess we could try harder to use the type system and compiler here... I'm fine with this, though.

bensigelman commented on opentracing-api/src/main/java/io/opentracing/Span.java:
```diff
@@ -19,60 +19,53 @@
-   * undefined behavior.
-   */
-  void finish();
+    /**
```
(indenting was off throughout this file)

bensigelman commented on opentracing-api/src/main/java/io/opentracing/Span.java:
```diff
@@ -19,60 +19,53 @@
-   *
-   * Note that newly-set Baggage items are only guaranteed to propagate to future children of the given Span.
-   */
-  Span setBaggageItem(String key, String value);
```
(removed / moved to SpanContext)

bensigelman commented on opentracing-api/src/main/java/io/opentracing/propagation/BinaryWriter.java:
```diff
@@ -1,19 +0,0 @@
-package io.opentracing.propagation;
-
-/** work in progress */
-public interface BinaryWriter {
```
no longer needed, correct? (if we rely on `ByteBuffer`)

adriancole commented on opentracing-api/src/main/java/io/opentracing/BuiltinFormats.java:
```diff
@@ -0,0 +1,44 @@
+
+/**
+ * BuiltinFormats encapsulates the inject()/extract() carrier formats that all Tracer implementations must support.
+ *
```
probably worth it to add docs for these generic params.. like..

```java
 * @param <R> foo used by {@link foo#bar} to bar things
 * @param <W> baz used by {@link baz#qux} to qux things
```

adriancole commented on opentracing-api/src/main/java/io/opentracing/BuiltinFormats.java:
```diff
@@ -0,0 +1,44 @@
+    /**
+     * BINARY carriers represent SpanContext state as an opaque byte array controlled entirely by the Tracer implementation.
+     */
+    public final static Format<ByteBuffer, ByteBuffer> BINARY = new BuiltinFormats<>();
```
bytebuffer is a fairly crap api (requires book-keeping, flip etc). since these are likely to be small, seems more straight-forward to use `byte[]`

michaelsembwever replied:
I'd really like to see example (s)of a ByteBuffer carrier in play.

My first attempt at implementing one raised a number of questions as to what usefulness any such defined Format and abstract Carrier can actually provide. (My efforts weren't very exhaustive so I won't be surprised to be proven wrong.)

Otherwise I would rather see ByteBuffer used than byte[]. (Sorry @adriancole )

adriancole replied:
>
> Otherwise I would rather see ByteBuffer used than byte[]. (Sorry
> @adriancole <https://github.com/adriancole> )
>
ByteBuffer is a crap api, but it is in-the-box and a few libraries use it
natively. Plus you can wrap byte arrays in with ByteBuffer etc. If we have
a compelling case that needed to represent bytes, I'm ok with ByteBuffer
(whine noted)

bensigelman replied:
I didn't see any great options here but don't feel strongly... I ended up going with `ByteBuffer` :-/


In other languages we have learned that it's often beneficial for the carrier to be an interface describing the reading/writing of data rather than being an actual data structure like ByteBuffer or byte[]. This leads me to the only other defensible (and in-the-box) choice: OutputStream/InputStream. Thoughts on that from people who use Java more than I do? (i.e., anyone :)

yurishkuro replied:
> OutputStream/InputStream

Not a bad choice, but leaves certain questions like "do I have to call `close()` on it when done?"

nio channels have quite minimal API surface, but ironically they do take ByteBuffer(), e.g. [WriteableByteChannel](https://docs.oracle.com/javase/7/docs/api/java/nio/channels/WritableByteChannel.html). And same problem with `close()`.

My Java experience has always been much higher in the stack, so I don't have a strong opinion.

adriancole commented on opentracing-api/src/main/java/io/opentracing/Reference.java:
```diff
@@ -0,0 +1,65 @@
+ between Spans.
+ */
+public class Reference {
+    private Type type;
```
nit: final classes and field references unless something is intended to be extended, mutated

ex.

```java
public final class Foo {
  private final Type type;

```

michaelsembwever replied:
+1

bensigelman replied:
thank you... I <3 nits and would like to be a better Java programmer :)

adriancole commented on opentracing-api/src/main/java/io/opentracing/Reference.java:
```diff
@@ -0,0 +1,65 @@
+    /**
+     * @return the SpanContext referred to by this reference.
+     */
+    public SpanContext referee() { return this.referee; }
```
referee is a distracting term. it is usually used in sports

michaelsembwever replied:
+1

bensigelman replied:
what's wrong with sports??? 🏈  🏉 

I changed this. I agree it's a little too cute / wonky.

adriancole commented on opentracing-api/src/main/java/io/opentracing/Tracer.java:
```diff
@@ -20,73 +20,80 @@
+   * @see BuiltinFormats
    */
-  <T> void inject(Span span, T carrier);
+  <W> void inject(SpanContext spanContext, Format<?, W> format, W carrier);
```
this Format type doesn't seem to be buying us much, as it is only used one-sided. The result looks more awkward than before to me.

adriancole replied:
the goal seems to be to top-level the difference between writing strings for use in http vs not http. Seems a penalty to have the standard allow both as it forces us to have an api to express a qualifier of a type.

Since `TextMapReader` is a contrived type anyway, how about `HttpHeaderReader`? Then you don't need to special-case and then add this Format thing, which would unravel some of the awkwardness.

yurishkuro replied:
`HttpHeaderReader` is also a qualifier of type, only (a) we need two of those, reader and writer, instead of just one new Format, and (b) it mixes two things, _how_ the data is given to the carrier, and _what_ data can be given.

Cf. https://github.com/opentracing/opentracing-java/issues/31#issuecomment-232394469

bensigelman replied:
@adriancole re TextMapReader and HttpHeaderReader: the pattern that's emerging is that some systems (e.g., gRPC) support unrestricted, sane, key:value string metadata maps. And then HTTP is a total mess from a key-mangling standpoint... yet it's obviously vitally important.

So, the gradual consensus has been that HTTP is "special" in that OpenTracing will recognize it with its own carrier (or pair of reader/writer carriers, whatever). But that we should also require Tracers to implement something less annoying like TextMap.

PS: none of this is at all interesting or even relevant if OT users aren't using Baggage.

adriancole replied:
The Format thing, which accepts any input and uses `?` as one of the parameters is feigning type safety. There's nothing that binds R and W together when `?` is used, right?

I'm most concerned about how this api is designed, not what it is trying to solve.

I think if you posted to a larger audience, they would agree that this api is tedious, almost scala in java:

```java
<W> void inject(SpanContext spanContext, Format<?, W> format, W carrier);
```

If you need to express a qualifier, you can append it as a parameter, but I wouldn't use a dual-parameter qualifier key knowing one would be ignored. If the only use-case is http, and the only case is baggage, I'd at least consider finding another way.

Most importantly, when designing a new api we should cite other similar designs, especially when they are full of types.

For example, there's a similar concern in dependency injection. They `Key` type below is made to both hold a qualifier and also the type of the instance passed in, (in this case `W`). It has more responsibility than what we are trying to do, yet has a less complex signature. Let's try to simplify!

https://google.github.io/guice/api-docs/latest/javadoc/index.html?com/google/inject/Key.html

adriancole replied:
later comment comparing with Key in dependency injection (which store the
type of W and allow it to be qualified by a string, annotation type etc).
basically, my biggest complaint is with the extra types in Format which are
ignored. Also the lack of a basic api (ex that whatever format becomes is a
mandatory parameter).

michaelsembwever replied:
In addition to @adriancole's concerns around the suitability of this api for java, i'm wondering if we should be adding an inject method rather than replacing the inject method.

That is to have both methods:

    <T> void inject(SpanContext span, T carrier);
    <W> void inject(SpanContext spanContext, …something more complex…);

Maintaining the simple approach for those use-cases where multiple mappings on carriers is not needed.

adriancole replied:
+1 wires crossed, but definitely agree on keeping the old one.

Most apis that i enjoy using only require advanced apis when you are doing
something advanced.

http://square.github.io/moshi/1.x/moshi/

yurishkuro replied:
Keeping both would've been a good compromise if not for the proposed `Format.HTTPHeader` that actually requires the 2-argument version, and is _the_ prevalent use case.

adriancole replied:
depends on how we document it, right?

keeping both could mean defer the decision to impl, or in the case of text
say it will use Format.HTTPHeader, etc

yurishkuro replied:
not sure if documenting will help. The point is that for http header we still want to use the same `TextMapReader/Writer` interface as for unrestricted maps, so we either differentiate by a Format constant, or by duplicating `TextMapReader/Writer`. Duplicating the interfaces just to force a different branch in the tracer feels an anti-pattern to me, quite worse than passing a second Format.X constant as differentiator.

adriancole replied:
I don't expect that you would choose to use the other method, yuri. I'm
speaking about users who are farther away from internals than you are, who
may prefer simpler choices.

yurishkuro replied:
Users, exactly. Right now the users see this method:

```java
void <T> inject(s SpanContext, T carrier)
``` 

Since `T` is unconstrained in any way, it's equivalent to 

```java
void inject(s SpanContext, Object carrier)
``` 

As a user, what am I supposed to pass as `carrier`?

Now consider the case with Format:

```java
void <W> inject(s SpanContext, Format<?, W> format, W carrier)
``` 

What would the user do? They will look at Format class and see a constant: 

```java
Format<.., TextMapWriter> TextMap = ...
``

Now they know what to pass as the `format` parameter, and it explicitly constraints what they should pass as a `carrier`:

```java
tracer.inject(span.getContext(), Format.TextMap, new TextMapWriter(){...})
```

michaelsembwever replied:
It's a good argument but I think it's simplifying things a bit.

To use the inject method you need to understand what you're doing in relation to the wire and tracing headers. For example that such headers go into the httpHeaders becomes base knowledge.

The use-case of an application having to support multiple formats that use the same carrier i would guess is more rare. As applications become smaller in a microservices world. When each application specifically only registering/configuring the carriers it needs to use simplifies this further.

adriancole commented on opentracing-api/src/main/java/io/opentracing/Tracer.java:
```diff
@@ -20,73 +20,80 @@
+      /**
+       * A shorthand for withReference(Reference.childOf(parent)).
        */
-      SpanBuilder withOperationName(String operationName);
```
bye operation name.. you will be missed :P

bensigelman replied:
still there, of course, when creating the builder... but another nice thing about the move towards SpanContext (and extract instead of join) is that you basically never need to create a span when you don't already know the operation name.

adriancole replied:
cool thx

adriancole commented on opentracing-api/src/main/java/io/opentracing/Reference.java:
```diff
@@ -0,0 +1,65 @@
+ * References are used by Tracer.buildSpan() to describe the relationships
+ between Spans.
+ */
+public class Reference {
```
I don't quite understand why we are pulling out this type. Seems simpler to rename Reference.Type to ReferenceType and squash the rest of this class. We could make one of the referenceTypes default.

If there's some compelling reason for making this object, it is likely a good thing for the doc section.

bensigelman replied:
this is a good point, @adriancole... it made sense in other languages that aren't using the builder pattern, but for Java we only need the reference type. I'll push my version of this in a moment.

adriancole commented on opentracing-api/src/main/java/io/opentracing/Tracer.java:
```diff
@@ -20,73 +20,80 @@
 
     Span http = tracer.buildSpan("HandleHTTPRequest")
-                      .withParent(feed)
+                      .withChildOf(parentSpan.context())
```
with prefix isn't something I'm in love with. I'd optimize for this reading nicer. ex `asChildOf()` or `withContext(context, referenceType)`

michaelsembwever replied:
`asChildOf(..)` is nice.

adriancole commented on opentracing-api/src/main/java/io/opentracing/Tracer.java:
```diff
@@ -20,73 +20,80 @@
+       * Add a reference from the Span being built to a distinct (usually parent) Span. May be called multiple times to represent multiple such References.
        */
-      SpanBuilder withParent(Span parent);
+      SpanBuilder withReference(Reference ref);
```
if this can be called multiple times, we should use `add` prefix to make it more obvious

michaelsembwever commented:
Without the Format stuff from https://github.com/opentracing/opentracing-java/issues/31 i'm +1 on merging.

With the Format stuff it's currently a -1 from me, unless further discussion helps me see the light.

yurishkuro commented on opentracing-api/src/main/java/io/opentracing/Reference.java:
```diff
@@ -0,0 +1,65 @@
+    private Type type;
+    private SpanContext referee;
+
+    public enum Type {
```
I am ok with this being an enum, as a list of _build-in_ reference types, but the `type` field in the `Reference` class needs to be of comparable type, not this enum. Otherwise this is not extensible without OT API release.

bensigelman replied:
we talked about this briefly off github -- I've gone with `Comparable` where reference types are needed and am using Strings for the builtin ones.

yurishkuro commented on opentracing-api/src/main/java/io/opentracing/Format.java:
```diff
@@ -11,9 +11,16 @@
+ * @see io.opentracing.Tracer#extract(Format, Object)
+ */
+public interface Format<R, W> {
 }
```
I suggest adding `bool equals(Format<?, ?> other)` to the interface to force implementations to provide proper `equals` method, since `==` should not be used for formats.

bensigelman commented:
Meta-comment: I'm going to try and split the (controversial) `Format` stuff out of this PR inasmuch as that's possible. We can then talk about the SpanContext and Reference aspects in isolation. I'll try to make those changes in the next 4 hours or so.

bensigelman commented:
Thanks so much for the comments, all, PTAL.

adriancole commented:
my non-binding opinion is that the type complexity of `Format<R, W>` isn't worth putting into the bottom api. It would end up being wrapped I think.

Complex apis in java usually breed libraries that hide their complexity. If we want to avoid some side-project called OpenTracing-Lite, I'd suggest highly that we simplify as much as possible the front-door of this api.

adriancole commented:
good idea

adriancole commented on opentracing-api/src/main/java/io/opentracing/Tracer.java:
```diff
@@ -15,78 +15,86 @@
+       * @see io.opentracing.References
        */
-      SpanBuilder withParent(Span parent);
+      SpanBuilder addReference(Comparable referenceType, SpanContext referencedContext);
```
if using Comparable, we should provide rationale for the choice, for example, why ordering matters. It is distracting otherwise.

A guess is that we are expecting an implementation to have multiple parents and need to know the parent vs adopted vs step-parent? I suppose we are expecting the implementation to retain the order?

Another note, Comparable has a type parameter `<T>` which would lead to this method having a compiler warning.

bensigelman replied:
Yeah, I'll just change to `Object`. I was responding to an earlier comment about the downside of using the (now obliterated) `Reference.Type` enum, as enums by definition are not externally extensible. I don't think the use case for a global ordering nearly compelling enough... we really just want equality comparisons.

adriancole replied:
:thumbsup:

adriancole commented on opentracing-api/src/main/java/io/opentracing/propagation/HttpHeaderReader.java:
```diff
@@ -0,0 +1,39 @@
+     *
+     * Note that these headers will often be a superset of whatever was injected via an HttpHeaderWriter in the peer. As such, Tracers should use a unique prefix or substring to identify their header map entries.
+     *
+     * @param key a key suitable for use in an HTTP header (i.e., case-insensitive, no special characters, etc)
```
should highlight that keys can repeat (since headers can have many values)

bensigelman replied:
also, this comment was copy-paste detritus... there should only be a `@return` here

adriancole commented on opentracing-api/src/main/java/io/opentracing/propagation/HttpHeaderReader.java:
```diff
@@ -0,0 +1,39 @@
+     *
+     * @see io.opentracing.Tracer#extract(Object)
+     */
+    Iterable<Map.Entry<String,String>> getEntries();
```
there's another choice, which is to return a multimap. Very often the impl will need to copy-out anyway..

Ex from okhttp:

```java
  public Map<String, List<String>> toMultimap() {
    Map<String, List<String>> result = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
    for (int i = 0, size = size(); i < size; i++) {
      String name = name(i).toLowerCase(Locale.US);
      List<String> values = result.get(name);
      if (values == null) {
        values = new ArrayList<>(2);
        result.put(name, values);
      }
      values.add(value(i));
    }
    return result;
  }
```

bensigelman replied:
I prefer the `Iterable` here since it's more general and allows for fewer copies in some circumstances... good point about the comment above, though, I'll change that.

adriancole replied:
unless the user needs to rewind, Iterator is slightly less hassle to
implement.

adriancole replied:
s/rewind/replay

bensigelman replied:
ah, good call -- changed.

adriancole commented on opentracing-api/src/main/java/io/opentracing/propagation/HttpHeaderWriter.java:
```diff
@@ -0,0 +1,35 @@
+     *
+     * @see io.opentracing.Tracer#inject(io.opentracing.SpanContext, Object)
+     */
+    void put(String key, String value);
```
since http headers can repeat, should we define what should occur when the key exists?

bensigelman replied:
this is just me, but I'd rather not. HTTP header support is so inconsistent across a large system, and IMO it's bad practice for a tracing system to rely on repeated headers anyway...

adriancole replied:
I agree I don't think we should support duplicate values, just curious if
we should ask implementors to prevent this somehow. Not documenting is fine
as we could later if it became common enough concern.

bensigelman replied:
👍 

In practice there will be O(1) headers for things like trace/span ids, and then zero or more headers for baggage. Since the baggage doesn't allow for duplicate keys, this is unlikely to become an issue at this layer (or so I hope).

tinkerware commented on opentracing-api/src/main/java/io/opentracing/Tracer.java:
```diff
@@ -15,78 +15,86 @@
+       * @see io.opentracing.References
        */
-      SpanBuilder withParent(Span parent);
+      SpanBuilder addReference(Object referenceType, SpanContext referencedContext);
```
I understand reference types need to be extensible, but is there a need to make them arbitrary types? Why not `String`?

bensigelman replied:
If all we need is `equals()`, I guess I would ask *why* String?

tinkerware replied:
If we had a `Symbol` type in Java, it would be that. Failing that, isn't it better that we at least pick something that  can `switch`'ed on?

adriancole replied:
good point on switching. enum couldn't be used bc enums aren't extensible.
for switch, we're left with ints or strings, strings require java 7+ (which
I think this library does anyway)

bensigelman replied:
this is a String now.

bensigelman commented:
@adriancole @michaelsembwever @yurishkuro @tinkerware

Since the (controversial) `Format` stuff has been split off from this PR, are there any objections to the remaining content? It seemed like we'd converged on the SpanContext / Reference / extract() aspects presented here.

I'm presently fixing the rest of the (github) repo to work along with these changes... it does require some surgery here and there but most of the diff will be uncontroversial. It's also far less important since OT callers don't depend on any of the impl helpers directly, of course.

Setting expectations: I'd like to merge this PR in 24-36h or so unless substantial new issues are raised.

tinkerware commented on opentracing-api/src/main/java/io/opentracing/Tracer.java:
```diff
@@ -15,78 +15,86 @@
+   * @see io.opentracing.propagation
    */
-  <T> void inject(Span span, T carrier);
+  <C> void inject(SpanContext spanContext, C carrier);
```
What's the point to the `C` type parameter when it does not constrain anything? If we are saying that a more elaborate format API is later work, we can just leave the carrier type `Object` for now, though it seems to me that there could at least be a marker interface for carriers.

yurishkuro replied:
I agree, this `C` makes it look like there's more to the `carrier` type than just an `Object`. +1 to drop it.

bensigelman replied:
This was inherited from the code at master...

https://github.com/opentracing/opentracing-java/blob/master/opentracing-api/src/main/java/io/opentracing/Tracer.java#L53

Aside from making the documentation more explicit (`@param <C> ...`), I can't provide any particular justification. I think this is code @michaelsembwever wrote, so I might be missing part of the rationale. (The documentation aspect is not trivial, though...)

michaelsembwever replied:
It helps implementations. See AbstractTracer for example.
It also creates a stub to explain Carrier types.

No biggie either way for me.

adriancole replied:
good point. I think that parameter doesn't make sense here since it doesn't
bind any other parameters or the return type.

tinkerware commented:
This seems okay to me, assuming that carrier/format API  is not in its final form.

yurishkuro commented on opentracing-api/src/main/java/io/opentracing/Tracer.java:
```diff
@@ -15,78 +15,86 @@
 
     Span http = tracer.buildSpan("HandleHTTPRequest")
-                      .withParent(feed)
+                      .asChildOf(parentSpan.context())
```
@bensigelman similar to Node/Python, we can overload this to `asChildOf(parentSpan)`

yurishkuro commented on opentracing-api/src/main/java/io/opentracing/References.java:
```diff
@@ -0,0 +1,35 @@
+ *
+ * References are used by Tracer.buildSpan() to describe the relationships between Spans.
+ *
+ * @see io.opentracing.Tracer.SpanBuilder#addReference(Comparable, SpanContext)
```
rm Comparable (a couple places)

yurishkuro commented:
lgtm

michaelsembwever commented:
:+1:

dawallin commented on opentracing-api/src/main/java/io/opentracing/SpanContext.java:
```diff
@@ -0,0 +1,39 @@
+     * @return the value of the baggage item identified by the given key, or null if no such item could be found
+     */
+    String getBaggageItem(String key);
+}
```
In go a method to iterate through all baggage items were introduced. Should you have it here as well ?

bensigelman replied:
I would like this to happen in a followup... there are some subtleties that @tinkerware brought up re immutability that I think are relevant here (namely, if we know that SpanContexts are immutable, the interface can be zero-copy *and* return an iterator *and* be threadsafe... otherwise I'd rather do something like what we did in Go and have the SpanContext impl invoke a callback repeatedly).

dawallin commented on opentracing-api/src/main/java/io/opentracing/SpanContext.java:
```diff
@@ -0,0 +1,39 @@
+     *
+     * @return this SpanContext instance, for chaining
+     */
+    SpanContext setBaggageItem(String key, String value);
```
I don't like the setBaggageItem method on the span context. It forces span context to be mutable and opened for side effects. I rather like to leave it on the aggregate root (the Span) which can forward the baggage item to the context internally. That would make it possible to have the spanContext immutable.

yurishkuro replied:
@dawallin if the SpanContext does not contain the baggage, then `tracer.Extract` cannot return just the SpanContext, and we're back to the problem with `Join` which was both extracting the context _and_ starting the span

dawallin replied:
The SpanContext should definitely contain the baggage, it is only the set method that I don't like since it makes the SpanContext mutable. The Extract method creates the SpanContext and can pass the baggage as a list in the spanContext constructor, it does not need the set method.

bensigelman replied:
@dawallin I see your point that the baggage could be set when the SpanContext impl is created. My best argument for keeping setBaggageItem on SpanContext is just symmetry: having the setter and getter on different objects will confuse people. There's also the question of whether baggage should be settable after Span `finish` time... if not, that would be yet another argument in favor of what you're proposing.

All of this said, I would like to keep the setter and getter where they are for this PR. Your arguments apply across languages and we should be discussing them in the right place, not here on ot-java. I've created https://github.com/opentracing/opentracing.github.io/issues/106 .

bensigelman commented:
Okay -- I changed a bunch of stuff and fixed the build and tests in the rest of this repo (including NoopTracer, which is probably the only part anyone actually uses at this point).

bensigelman commented on opentracing-api/src/main/java/io/opentracing/Span.java:
```diff
@@ -19,60 +19,53 @@
+     *
+     * @return the SpanContext that encapsulates Span state that should propagate across process boundaries.
+     */
+    SpanContext getContext();
```
I changed this to `getContext()` since this is Java :)

(I'd internalized Go's no-get getter convention, I guess)

adriancole replied:
going without a getX is not abnormal when there is no setter. A lot of
libraries follow this, including zipkin. I prefer is as context()

adriancole replied:
ps the history of getX setX is JavaBeans. this isn't a java bean, so
doesn't need to behold to that. A lot of the api has more to do with
builders than beans
http://www.informit.com/articles/article.aspx?p=1216151&seqNum=2

bensigelman commented:
(and if there are no further concerns, I will merge this sometime tomorrow)

bensigelman commented:
@adriancole I dropped the `getContext()` commit. Thanks for the history lesson, too. :)

bensigelman commented:
Okay, as promised, I'm merging this now. Thanks for all of the feedback.

https://github.com/opentracing/opentracing-java/issues/31 and https://github.com/opentracing/opentracing.github.io/issues/106 are still open and could impact this API going forward.

adriancole commented:
Thanks @bensigelman!
